File: deploy/team-child-automation/email-code.mjs
const PROVIDER_ORIGIN = 'https://ic.g-c.cc'
const DEFAULT_TIMEOUT_MS = 60_000
const DEFAULT_INTERVAL_MS = 5_000
const MAX_RESPONSE_BYTES = 256 * 1024

function emailCodeError(code) {
  return Object.assign(new Error(code), { code })
}

function validateLogin(email, token) {
  if (typeof email !== 'string' || email.length > 254 || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)) {
    throw emailCodeError('email_code_access_denied')
  }
  if (typeof token !== 'string' || !/^[A-Fa-f0-9]{64}$/.test(token)) {
    throw emailCodeError('email_code_access_denied')
  }
}

async function boundedText(response) {
  const length = Number(response.headers?.get?.('content-length') || 0)
  if (Number.isFinite(length) && length > MAX_RESPONSE_BYTES) throw emailCodeError('email_code_unavailable')
  const body = await response.text()
  if (Buffer.byteLength(body) > MAX_RESPONSE_BYTES) throw emailCodeError('email_code_unavailable')
  return body
}

async function request(fetchImpl, url, options = {}) {
  let response
  try {
    response = await fetchImpl(url, { redirect: 'error', ...options })
  } catch (error) {
    if (options.signal?.aborted || error?.name === 'AbortError') throw emailCodeError('email_code_canceled')
    throw emailCodeError('email_code_unavailable')
  }
  let body
  try {
    body = await boundedText(response)
  } catch (error) {
    if (options.signal?.aborted || error?.name === 'AbortError') throw emailCodeError('email_code_canceled')
    if (error?.code) throw error
    throw emailCodeError('email_code_unavailable')
  }
  let payload
  try {
    payload = body ? JSON.parse(body) : null
  } catch {
    throw emailCodeError('email_code_unavailable')
  }
  if (!response.ok || payload?.ok === false) {
    if (response.status === 401 || String(payload?.error?.code || '').toUpperCase() === 'UNAUTHORIZED') {
      throw emailCodeError('email_code_access_denied')
    }
    throw emailCodeError('email_code_unavailable')
  }
  return payload?.data
}

async function discoverGateway(fetchImpl, signal) {
  let response
  try {
    response = await fetchImpl(`${PROVIDER_ORIGIN}/`, {
      redirect: 'error',
      signal,
      headers: { Accept: 'text/html' },
    })
  } catch (error) {
    if (signal?.aborted || error?.name === 'AbortError') throw emailCodeError('email_code_canceled')
    throw emailCodeError('email_code_unavailable')
  }
  if (!response.ok) throw emailCodeError('email_code_unavailable')
  const html = await boundedText(response)
  const gateway = html.match(/data-public-mail-gateway=["'](\/m\/[A-Za-z0-9_-]{16,128})["']/i)?.[1] || ''
  if (!gateway) throw emailCodeError('email_code_unavailable')
  return gateway
}

function normalizeMessages(data) {
  return Array.isArray(data?.messages) ? data.messages.filter(item => item && typeof item === 'object').slice(0, 20) : []
}

function messageID(message) {
  return String(message?.id || '').trim()
}

function messageTimestamp(message) {
  const parsed = Date.parse(String(message?.date || ''))
  return Number.isFinite(parsed) ? parsed : 0
}

function expiryEpochSeconds(value) {
  const numeric = Number(value)
  if (Number.isFinite(numeric) && numeric > 0) return numeric > 1_000_000_000_000 ? numeric / 1000 : numeric
  const parsed = Date.parse(String(value || ''))
  return Number.isFinite(parsed) ? parsed / 1000 : 0
}

function readableHTML(value) {
  return String(value || '').replace(/<style\b[^>]*>[\s\S]*?<\/style>/gi, ' ')
    .replace(/<script\b[^>]*>[\s\S]*?<\/script>/gi, ' ')
    .replace(/<[^>]+>/g, ' ')
}

function messageText(message) {
  return [message?.from, message?.subject, message?.snippet, message?.text, message?.htmlPreview, readableHTML(message?.html)]
    .map(value => String(value || ''))
    .join('\n')
    .replace(/\s+/g, ' ')
    .trim()
}

function isOpenAIMessage(message) {
  return /\bopenai\b|\bchatgpt\b/i.test(messageText(message))
}

export function extractOpenAIVerificationCode(message) {
  const source = messageText(message)
  if (!source || !isOpenAIMessage(message)) return ''
  const contextual = source.match(/(?:verification|verify|login|sign[- ]?in|security|temporary|one[- ]?time|验证码|登录代码|临时代码)[^\d]{0,80}(\d{6})/i)
  if (contextual) return contextual[1]
  const candidates = [...new Set(Array.from(source.matchAll(/\b\d{6}\b/g), match => match[0]))]
  return candidates.length === 1 ? candidates[0] : ''
}

async function wait(ms, signal, setTimeoutImpl, clearTimeoutImpl) {
  if (signal?.aborted) throw emailCodeError('email_code_canceled')
  await new Promise((resolve, reject) => {
    let timer
    const cleanup = () => signal?.removeEventListener?.('abort', cancel)
    const cancel = () => {
      clearTimeoutImpl(timer)
      cleanup()
      reject(emailCodeError('email_code_canceled'))
    }
    const finish = () => {
      cleanup()
      resolve()
    }
    timer = setTimeoutImpl(finish, ms)
    signal?.addEventListener('abort', cancel, { once: true })
  })
}

export async function createICGCEmailCodeSession({ email, token, signal }, dependencies = {}) {
  validateLogin(email, token)
  const fetchImpl = dependencies.fetch || globalThis.fetch
  if (typeof fetchImpl !== 'function') throw emailCodeError('email_code_unavailable')
  const now = dependencies.now || Date.now
  const setTimeoutImpl = dependencies.setTimeout || setTimeout
  const clearTimeoutImpl = dependencies.clearTimeout || clearTimeout
  const normalizedEmail = email.trim().toLowerCase()
  let privateToken = token
  let gateway = await discoverGateway(fetchImpl, signal)
  let proof = ''
  let proofExpiresAt = 0
  let closed = false

  async function authorize() {
    const data = await request(fetchImpl, `${PROVIDER_ORIGIN}${gateway}/authorize`, {
      method: 'POST',
      signal,
      headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
      body: JSON.stringify({ email: normalizedEmail, token: privateToken, turnstileToken: '' }),
    })
    proof = String(data?.accessProof || '')
    proofExpiresAt = expiryEpochSeconds(data?.accessProofExpiresAt)
    if (!proof || proof.length > 4096) throw emailCodeError('email_code_unavailable')
  }

  function authHeaders() {
    return {
      Accept: 'application/json',
      'X-Public-Email': normalizedEmail,
      'X-Public-Email-Token': privateToken,
      'X-Mail-Access-Proof': proof,
    }
  }

  async function listMessages() {
    if (!proof || proofExpiresAt <= now() / 1000 + 5) await authorize()
    try {
      return normalizeMessages(await request(fetchImpl, `${PROVIDER_ORIGIN}${gateway}/items?limit=10`, {
        signal,
        headers: authHeaders(),
      }))
    } catch (error) {
      if (error?.code !== 'email_code_access_denied') throw error
      proof = ''
      await authorize()
      return normalizeMessages(await request(fetchImpl, `${PROVIDER_ORIGIN}${gateway}/items?limit=10`, {
        signal,
        headers: authHeaders(),
      }))
    }
  }

  async function messageDetail(id) {
    const data = await request(fetchImpl, `${PROVIDER_ORIGIN}${gateway}/items/${encodeURIComponent(id)}`, {
      signal,
      headers: authHeaders(),
    })
    return data?.message && typeof data.message === 'object' ? data.message : {}
  }

  await authorize()
  const baseline = new Set()
  for (const message of await listMessages()) {
    const id = messageID(message)
    if (id) baseline.add(id)
  }
  const processed = new Set(baseline)

  return {
    async waitForCode({ notBefore = now(), timeoutMs = DEFAULT_TIMEOUT_MS, intervalMs = DEFAULT_INTERVAL_MS } = {}) {
      const deadline = now() + timeoutMs
      let successfulQuery = false
      while (!closed && !signal?.aborted && now() < deadline) {
        try {
          const messages = await listMessages()
          successfulQuery = true
          messages.sort((left, right) => messageTimestamp(right) - messageTimestamp(left))
          for (const message of messages) {
            const id = messageID(message)
            if (!id || processed.has(id)) continue
            if (messageTimestamp(message) && messageTimestamp(message) < notBefore - 10_000) {
              processed.add(id)
              continue
            }
            if (!isOpenAIMessage(message)) {
              processed.add(id)
              continue
            }
            let code = extractOpenAIVerificationCode(message)
            if (!code) {
              try {
                code = extractOpenAIVerificationCode({ ...message, ...(await messageDetail(id)) })
              } catch (error) {
                if (error?.code === 'email_code_access_denied' || error?.code === 'email_code_canceled') throw error
                continue
              }
            }
            processed.add(id)
            if (code) return code
          }
        } catch (error) {
          if (error?.code === 'email_code_access_denied' || error?.code === 'email_code_canceled') throw error
        }
        const remaining = deadline - now()
        if (remaining <= 0) break
        await wait(Math.min(intervalMs, remaining), signal, setTimeoutImpl, clearTimeoutImpl)
      }
      if (closed || signal?.aborted) throw emailCodeError('email_code_canceled')
      throw emailCodeError(successfulQuery ? 'email_code_timeout' : 'email_code_unavailable')
    },
    close() {
      closed = true
      privateToken = ''
      proof = ''
      proofExpiresAt = 0
      baseline.clear()
      processed.clear()
      gateway = ''
    },
  }
}

export { PROVIDER_ORIGIN as ICGC_EMAIL_CODE_ORIGIN }

File: tools/xiass-adspower-helper/email_code.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"html"
	"io"
	"log"
	"net/http"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

const emailCodeProviderOrigin = "https://ic.g-c.cc"

var (
	emailCodeGatewayPattern   = regexp.MustCompile(`(?i)data-public-mail-gateway=["'](/m/[A-Za-z0-9_-]{16,128})["']`)
	emailCodeHTMLPattern      = regexp.MustCompile(`(?s)<[^>]+>`)
	emailCodeContextPattern   = regexp.MustCompile(`(?i)(?:verification|verify|login|sign[- ]?in|security|temporary|one[- ]?time|验证码|登录代码|临时代码)[^0-9]{0,80}([0-9]{6})`)
	emailCodeSixDigitsPattern = regexp.MustCompile(`\b[0-9]{6}\b`)
)

type emailCodeError string

func (e emailCodeError) Error() string { return string(e) }

type emailCodeSession struct {
	client     *http.Client
	email      string
	token      string
	gateway    string
	proof      string
	proofUntil time.Time
	baseline   map[string]bool
}

type emailCodeMessage struct {
	ID          string `json:"id"`
	Date        string `json:"date"`
	From        string `json:"from"`
	Subject     string `json:"subject"`
	Snippet     string `json:"snippet"`
	Text        string `json:"text"`
	HTML        string `json:"html"`
	HTMLPreview string `json:"htmlPreview"`
}

func newEmailCodeSession(ctx context.Context, address, token string) (*emailCodeSession, error) {
	address = strings.ToLower(strings.TrimSpace(address))
	token = strings.TrimSpace(token)
	if len(address) > 254 || !strings.Contains(address, "@") || len(token) != 64 {
		return nil, emailCodeError("email_code_access_denied")
	}
	for _, char := range token {
		if !strings.ContainsRune("0123456789abcdefABCDEF", char) {
			return nil, emailCodeError("email_code_access_denied")
		}
	}
	session := &emailCodeSession{
		client: &http.Client{Timeout: 20 * time.Second, CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }},
		email:  address, token: token, baseline: make(map[string]bool),
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, emailCodeProviderOrigin+"/", nil)
	if err != nil {
		return nil, emailCodeError("email_code_unavailable")
	}
	request.Header.Set("Accept", "text/html")
	response, err := session.client.Do(request)
	if err != nil || response.StatusCode != http.StatusOK {
		if response != nil {
			_ = response.Body.Close()
		}
		return nil, emailCodeError("email_code_unavailable")
	}
	body, err := readBounded(response.Body, 256<<10)
	_ = response.Body.Close()
	if err != nil {
		return nil, emailCodeError("email_code_unavailable")
	}
	match := emailCodeGatewayPattern.FindSubmatch(body)
	if len(match) != 2 {
		return nil, emailCodeError("email_code_unavailable")
	}
	session.gateway = string(match[1])
	if err := session.authorize(ctx); err != nil {
		return nil, err
	}
	messages, err := session.list(ctx)
	if err != nil {
		return nil, err
	}
	for _, message := range messages {
		if message.ID != "" {
			session.baseline[message.ID] = true
		}
	}
	return session, nil
}

func readBounded(reader io.Reader, limit int64) ([]byte, error) {
	limited := io.LimitReader(reader, limit+1)
	payload, err := io.ReadAll(limited)
	if err != nil || int64(len(payload)) > limit {
		return nil, errors.New("response too large")
	}
	return payload, nil
}

func (s *emailCodeSession) requestJSON(ctx context.Context, method, path string, body any, headers map[string]string, target any) error {
	var reader io.Reader
	if body != nil {
		payload, err := json.Marshal(body)
		if err != nil {
			return emailCodeError("email_code_unavailable")
		}
		reader = bytes.NewReader(payload)
	}
	request, err := http.NewRequestWithContext(ctx, method, emailCodeProviderOrigin+s.gateway+path, reader)
	if err != nil {
		return emailCodeError("email_code_unavailable")
	}
	request.Header.Set("Accept", "application/json")
	if body != nil {
		request.Header.Set("Content-Type", "application/json")
	}
	for key, value := range headers {
		request.Header.Set(key, value)
	}
	response, err := s.client.Do(request)
	if err != nil {
		return emailCodeError("email_code_unavailable")
	}
	defer response.Body.Close()
	payload, err := readBounded(response.Body, 256<<10)
	if err != nil {
		return emailCodeError("email_code_unavailable")
	}
	if response.StatusCode == http.StatusUnauthorized {
		return emailCodeError("email_code_access_denied")
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return emailCodeError("email_code_unavailable")
	}
	var envelope struct {
		OK    *bool           `json:"ok"`
		Data  json.RawMessage `json:"data"`
		Error struct {
			Code string `json:"code"`
		} `json:"error"`
	}
	if json.Unmarshal(payload, &envelope) != nil || envelope.OK != nil && !*envelope.OK {
		if strings.EqualFold(envelope.Error.Code, "UNAUTHORIZED") {
			return emailCodeError("email_code_access_denied")
		}
		return emailCodeError("email_code_unavailable")
	}
	if target != nil && len(envelope.Data) > 0 && string(envelope.Data) != "null" {
		if json.Unmarshal(envelope.Data, target) != nil {
			return emailCodeError("email_code_unavailable")
		}
	}
	return nil
}

func (s *emailCodeSession) authorize(ctx context.Context) error {
	var data struct {
		AccessProof          string `json:"accessProof"`
		AccessProofExpiresAt any    `json:"accessProofExpiresAt"`
	}
	if err := s.requestJSON(ctx, http.MethodPost, "/authorize", map[string]string{
		"email": s.email, "token": s.token, "turnstileToken": "",
	}, nil, &data); err != nil {
		return err
	}
	s.proof = strings.TrimSpace(data.AccessProof)
	if s.proof == "" || len(s.proof) > 4096 {
		return emailCodeError("email_code_unavailable")
	}
	s.proofUntil = time.Now().Add(2 * time.Minute)
	switch value := data.AccessProofExpiresAt.(type) {
	case float64:
		if value > 1_000_000_000_000 {
			value /= 1000
		}
		s.proofUntil = time.Unix(int64(value), 0)
	case string:
		if parsed, err := time.Parse(time.RFC3339, value); err == nil {
			s.proofUntil = parsed
		}
	}
	return nil
}

func (s *emailCodeSession) authHeaders() map[string]string {
	return map[string]string{"X-Public-Email": s.email, "X-Public-Email-Token": s.token, "X-Mail-Access-Proof": s.proof}
}

func (s *emailCodeSession) list(ctx context.Context) ([]emailCodeMessage, error) {
	if s.proof == "" || time.Until(s.proofUntil) < 5*time.Second {
		if err := s.authorize(ctx); err != nil {
			return nil, err
		}
	}
	var data struct {
		Messages []emailCodeMessage `json:"messages"`
	}
	err := s.requestJSON(ctx, http.MethodGet, "/items?limit=10", nil, s.authHeaders(), &data)
	if errors.Is(err, emailCodeError("email_code_access_denied")) {
		s.proof = ""
		if authErr := s.authorize(ctx); authErr != nil {
			return nil, authErr
		}
		err = s.requestJSON(ctx, http.MethodGet, "/items?limit=10", nil, s.authHeaders(), &data)
	}
	return data.Messages, err
}

func (s *emailCodeSession) detail(ctx context.Context, id string) (emailCodeMessage, error) {
	var data struct {
		Message emailCodeMessage `json:"message"`
	}
	err := s.requestJSON(ctx, http.MethodGet, "/items/"+url.PathEscape(id), nil, s.authHeaders(), &data)
	return data.Message, err
}

func (s *emailCodeSession) waitForCode(ctx context.Context, notBefore time.Time) (string, error) {
	timer := time.NewTimer(10 * time.Second)
	select {
	case <-ctx.Done():
		timer.Stop()
		return "", emailCodeError("email_code_timeout")
	case <-timer.C:
	}
	deadline := time.Now().Add(2 * time.Minute)
	for time.Now().Before(deadline) {
		messages, err := s.list(ctx)
		if err != nil {
			return "", err
		}
		now := time.Now()
		candidate, ok := latestEmailCodeCandidate(messages, s.baseline, notBefore, now)
		if ok {
			if code := extractEmailCode(candidate); code != "" {
				log.Printf("email code candidate accepted: age=%s", now.Sub(parseMessageTime(candidate.Date)).Round(time.Second))
				return code, nil
			}
			// Do not fall back to an older message. The newest eligible message is
			// the only one that can belong to this authorization attempt. A detail
			// request may fail temporarily, so leave the candidate eligible for the
			// next poll instead of permanently marking it as processed.
			detail, detailErr := s.detail(ctx, candidate.ID)
			if detailErr == nil {
				if code := extractEmailCode(detail); code != "" {
					log.Printf("email code detail accepted: age=%s", now.Sub(parseMessageTime(candidate.Date)).Round(time.Second))
					return code, nil
				}
			}
		}
		timer := time.NewTimer(5 * time.Second)
		select {
		case <-ctx.Done():
			timer.Stop()
			return "", emailCodeError("email_code_timeout")
		case <-timer.C:
		}
	}
	return "", emailCodeError("email_code_timeout")
}

func latestEmailCodeCandidate(messages []emailCodeMessage, baseline map[string]bool, requestedAt, now time.Time) (emailCodeMessage, bool) {
	ordered := append([]emailCodeMessage(nil), messages...)
	sort.SliceStable(ordered, func(i, j int) bool {
		left, right := parseMessageTime(ordered[i].Date), parseMessageTime(ordered[j].Date)
		if left.IsZero() {
			return false
		}
		if right.IsZero() {
			return true
		}
		return left.After(right)
	})
	for _, message := range ordered {
		if message.ID == "" || baseline[message.ID] || !eligibleEmailCodeMessage(message, requestedAt, now) {
			continue
		}
		return message, true
	}
	return emailCodeMessage{}, false
}

func eligibleEmailCodeMessage(message emailCodeMessage, requestedAt, now time.Time) bool {
	messageAt := parseMessageTime(message.Date)
	if messageAt.IsZero() {
		return false
	}
	if requestedAt.IsZero() || messageAt.Before(requestedAt.Add(-5*time.Second)) {
		return false
	}
	if messageAt.After(now.Add(15 * time.Second)) {
		return false
	}
	return now.Sub(messageAt) <= time.Minute
}

func parseMessageTime(value string) time.Time {
	value = strings.TrimSpace(value)
	if value == "" {
		return time.Time{}
	}
	for _, layout := range []string{
		time.RFC3339Nano,
		time.RFC3339,
		time.RFC1123Z,
		time.RFC1123,
		time.RFC822Z,
		time.RFC822,
		"2006-01-02 15:04:05Z07:00",
		"2006-01-02 15:04:05",
	} {
		if parsed, err := time.Parse(layout, value); err == nil {
			return parsed
		}
	}
	if numeric, err := strconv.ParseInt(value, 10, 64); err == nil {
		if numeric > 1_000_000_000_000 {
			return time.UnixMilli(numeric)
		}
		if numeric > 0 {
			return time.Unix(numeric, 0)
		}
	}
	return time.Time{}
}

func messagePredates(value string, cutoff time.Time) bool {
	parsed := parseMessageTime(value)
	return !parsed.IsZero() && parsed.Before(cutoff)
}

func extractEmailCode(message emailCodeMessage) string {
	source := strings.Join([]string{message.From, message.Subject, message.Snippet, message.Text, message.HTMLPreview, emailCodeHTMLPattern.ReplaceAllString(html.UnescapeString(message.HTML), " ")}, " ")
	if !regexp.MustCompile(`(?i)openai|chatgpt`).MatchString(source) {
		return ""
	}
	if match := emailCodeContextPattern.FindStringSubmatch(source); len(match) == 2 {
		return match[1]
	}
	matches := emailCodeSixDigitsPattern.FindAllString(source, -1)
	unique := make(map[string]bool)
	for _, match := range matches {
		unique[match] = true
	}
	if len(unique) == 1 {
		for value := range unique {
			return value
		}
	}
	return ""
}

func emailCodeReason(err error) string {
	var coded emailCodeError
	if errors.As(err, &coded) {
		switch string(coded) {
		case "email_code_access_denied", "email_code_unavailable", "email_code_timeout":
			return string(coded)
		}
	}
	return "email_code_unavailable"
}

func (s *emailCodeSession) String() string {
	return fmt.Sprintf("email code session for %s", s.email)
}
